### PR TITLE
 [configure] Add -allow-unspecified-lablgtk flag 

### LINKED ISSUE #16047: Please add a `./configure` flag to suppress the lablgtk3 version check when the version is not a number
#### Description of the problem

CoqIDE fails to build on new versions of Ubuntu because [`ocamlfind query -format "%v" lablgtk3` reports `[unspecified]`](https://bugs.launchpad.net/ubuntu/+source/lablgtk3/+bug/1964437) (see also [this](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=971439) and [this](https://bugs.launchpad.net/ubuntu/+source/lablgtk3/+bug/1897171) and https://github.com/coq/coq/issues/13079).  I would appreciate a `./configure` flag that allows building CoqIDE anyway when the version of lablgtk3 is not recognized as a version number.

#### Coq Version

master, 8.15

### PR DESCRIPTION
**Pull request by Alizter**
Fix #16047

@JasonGross do you have a way of testing this?

Also do you want this in 8.15/8.16/8.17?

### CODE REVIEW

**Comment by JasonGross** [comment]
I think the flag should only ignore unknown versions, i.e., ones that land in the catch branch of the try?  And, yes, backporting would be quite nice, as far back as reasonable.

**Comment by Alizter** [comment]
Have you got an example of `ocamlfind` returning `UNDEFINED` or something? I can try and match the version on that.

**Comment by JasonGross** [comment]
I believe it returns the string `[unspecified]`, and you should probably be able to reproduce this by figuring out where the `META` file lives for your lablgtk and renaming it temporarily to `META.bak` (or move it to `/tmp`).  I guess maybe this gets parsed as the empty version list?

**Comment by ejgallego** [comment]
Maybe we could just have a flag that ignores all version checks? That could be useful for experiments.

File: tools/configure/coqide.ml [hunk 1]
```diff
@@ -58,7 +61,7 @@ let check_coqide ocamlfind prefs best_compiler camllib =
   if dir = ""
   then set_ide prefs No "LablGtk3 or LablGtkSourceView3 not found"
   else
-    let (ok, version) = check_lablgtk_version ocamlfind in
+    let (ok, version) = check_lablgtk_version prefs ocamlfind in
```
Comment (ejgallego) [inline tools/configure/coqide.ml:64]: Another choice pass directly `prefs.ignorelablgtk`

**Comment by Alizter** [comment]
@ejgallego Currently this ignores all version checks. Jason is asking if we can still disambiguate between no-lablgtk and an unknown lablgtk version. I will try to see what I can do.

**Comment by Alizter** [comment]
When I change my META to the versions described, ocamlfind does not struggle with the versions. I suspect that the META file in question actually contains "[unspecified]" as the version field.

**Comment by Alizter** [comment]
I've tested this my self on a META file with the version "[unspecified]". Without the flag configure will not intend to build CoqIDE as reported, with the flag it forgoes the check and intends to build anyway.

The flag is `-allow-unspecified-lablgtk`.

**Review by JasonGross — APPROVED** [review]
Thanks!  If others such as @ejgallego would prefer the "ignore lablgtk version check" flag instead of this one, I'm fine with that version instead.  The reason that I requested "ignore only unspecified version" is that I want this to be a relatively safe flag to pass; I'd prefer for my packages that pass this to still error if the lablgtk version is known to be too old, so that I catch that issue.

**Comment by JasonGross** [comment]
> I've tested this myself on a META file with the version "[unspecified]". Without the flag configure will not intend to build CoqIDE as reported, with the flag it forgoes the check and intends to build anyway.

Ah, you're right, according to [the file list](https://packages.ubuntu.com/kinetic/amd64/liblablgtk3-ocaml/filelist) there is a `META` file, and the contents seems to be
```
description = ""
requires = "cairo2 threads"
archive(byte) = "lablgtk3.cma"
archive(native) = "lablgtk3.cmxa"
plugin(byte) = "lablgtk3.cma"
plugin(native) = "lablgtk3.cmxs"
```

**Comment by Alizter** [comment]
cc @SkySkimmer can this make 8.15.2?

**Comment by ejgallego** [comment]
Something that feels weird to me that is that specifically add a flag to workaround a packaging bug. In the sense, I don't think there is a real use case for this flag except for workarounding buggy packaging. That seems wrong to me.

A flag that disables dep checking altoghether seems to have more uses to me.

**Comment by SkySkimmer** [comment]
Note that configure's coqide setting AFAIK only matters for this bit in makefile.ide
~~~make
ifeq ($(HASCOQIDE),no)
coqide:
	>&2 echo "CoqIDE compilation configured off."
	false
else
coqide: $(COQIDE) $(IDETOP) $(TOPBINOPT) $(IDEBINDINGS) $(VO_OUT_DIR)theories/Init/Prelude.$(VO) $(BCONTEXT)/coqide.install
endif
~~~
We could do instead like for the doc target
~~~make
coqide: coqide-$(HASCOQIDE)

coqide-no:
	>&2 echo "CoqIDE compilation configured off."
	false

coqide-yes: $(COQIDE) $(IDETOP) $(TOPBINOPT) $(IDEBINDINGS) $(VO_OUT_DIR)theories/Init/Prelude.$(VO) $(BCONTEXT)/coqide.install
~~~
then you could `make coqide-yes` to force compilation regardless of detected gtk

File: tools/configure/coqide.ml [hunk 2]
```diff
@@ -23,15 +23,19 @@ let get_lablgtkdir ocamlfind =
...
+let check_lablgtk_version ?(unspeclablgtk=false) ocamlfind =
   let v, _ = tryrun ocamlfind ["query"; "-format"; "%v"; "lablgtk3"] in
-  try
-    let vn = generic_version_nums ~name:"lablgtk3" v in
-    if vn < [3; 1; 2] then
-      (false, v)
-    else
-      (true, v)
-  with _ -> (false, v)
+  (* When the *)
```
Comment (ejgallego) [inline tools/configure/coqide.ml:28]: This comment seems undone?

**Comment by ejgallego** [comment]
This was discussed in the call, and indeed I'm not comfortable polluting configure's option and documentation space with this workaround for a Debian bug. We can either introduce a general flag --ignore-version-checks or Debian packages can just disable this check locally.

**Comment by Alizter** [comment]
if @JasonGross is happy submitting patches to debian, then I am happy to close this.

**Comment by SkySkimmer** [comment]
Don't need this in 8.15.2

**Comment by Alizter** [comment]
@ppedrot I've removed the milestone.

**Comment by Alizter** [comment]
Closing this since it doesn't look like we want to go this direction.

**Comment by ejgallego** [comment]
We don't really have a consistent policy on how much we support downstream distributions brokeness in Coq. It'd be good if we did.

**Pull request closed without merging**